src/vma/iomux/epfd_info.cpp

```diff
@@ -595,25 +606,26 @@ void epfd_info::insert_epoll_event_cb(int fd, uint32_t event_flags)
 
 void epfd_info::insert_epoll_event(int fd, uint32_t event_flags)
 {
-	ep_ready_fd_map_t::iterator iter = m_ready_fds.find(fd);
-	if (iter != m_ready_fds.end()) {
-		iter->second |= event_flags;
+	socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
+	if (sock_fd->ep_ready_fd_node.is_list_member()) {
+		sock_fd->m_epoll_event_flags |= event_flags;
 	}
 	else {
-		m_ready_fds.insert(ep_ready_fd_map_t::value_type(fd, event_flags));
+		sock_fd->m_epoll_event_flags = event_flags;
+		m_ready_fds.push_back(sock_fd);
 	}
 
 	do_wakeup();
 }
 
 void epfd_info::remove_epoll_event(int fd, uint32_t event_flags)
 {
-	ep_ready_fd_map_t::iterator iter = m_ready_fds.find(fd);
-	if (iter != m_ready_fds.end()) {
-		iter->second &= ~event_flags;
-	}
-	if (iter->second == 0) {
-		m_ready_fds.erase(iter);
+	socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
+	if (sock_fd->ep_ready_fd_node.is_list_member()) {
+		sock_fd->m_epoll_event_flags &= ~event_flags;
+		if (sock_fd->m_epoll_event_flags == 0) {
```

Comment:
The purpose of the second if is to remove the fd from m_ready_fds if the event_flags value is 0.
Like your comment above, we have to check if the fd is a list member before removing it from the list.

> 
socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
if (sock_fd->ep_ready_fd_node.is_list_member()) {
      sock_fd->m_epoll_event_flags &= ~event_flags;
}
if (sock_fd->ep_ready_fd_node.is_list_member() && sock_fd->m_epoll_event_flags == 0) {
m_ready_fds.erase(sock_fd);
}

This is equal to my implementation - insert the second if into the first if and remove the is_list_member() check from the if statement:

> 
socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
if (sock_fd->ep_ready_fd_node.is_list_member()) {
sock_fd->m_epoll_event_flags &= ~event_flags;
if (sock_fd->m_epoll_event_flags == 0) {
m_ready_fds.erase(sock_fd);
}
}